python-package/lightgbm/basic.py

```diff
@@ -1853,6 +1853,119 @@ def free_network(self):
+        """
+        if not PANDAS_INSTALLED:
+            warnings.warn('Pandas is not installed and so the result will be returned as a list')
+
+        if self.num_trees() == 0:
+            raise LightGBMError('There are no trees in this Booster and thus nothing to parse')
+
+        def _is_split_node(tree):
+            return 'split_index' in tree.keys()
+
+        def create_node_record(tree, node_depth=1, tree_index=None,
+                               feature_names=None, parent_node=None):
+
+            def _get_node_index(tree, tree_index):
+                tree_num = str(tree_index) + '-' if tree_index is not None else ''
+                is_split = _is_split_node(tree)
+                node_type = 'S' if is_split else 'L'
+                node_num = str(tree['split_index' if is_split else 'leaf_index'])
+                return tree_num + node_type + node_num
+
+            def _get_split_feature(tree, feature_names):
+                if _is_split_node(tree):
+                    if feature_names is not None:
+                        feature_name = feature_names[tree['split_feature']]
+                    else:
+                        feature_name = tree['split_feature']
+                else:
+                    feature_name = None
+                return feature_name
+
+            # Create the node record, and populate universal data members
+            node = {'tree_index': tree_index,
+                    'node_depth': node_depth,
+                    'node_index': _get_node_index(tree, tree_index),
+                    'left_child': None,
+                    'right_child': None,
+                    'parent_index': parent_node,
+                    'split_feature': _get_split_feature(tree, feature_names),
+                    'split_gain': None,
+                    'threshold': None,
+                    'decision_type': None,
+                    'missing_direction': None,
+                    'missing_type': None,
+                    'internal_value': None,
+                    'leaf_value': None,
+                    'weight': None,
+                    'count': None}
+
+            # Update values to reflect node type (leaf or split)
+            if _is_split_node(tree):
+                node['left_child'] = _get_node_index(tree['left_child'], tree_index)
+                node['right_child'] = _get_node_index(tree['right_child'], tree_index)
+                node['split_gain'] = tree['split_gain']
+                node['threshold'] = tree['threshold']
+                node['decision_type'] = tree['decision_type']
+                node['missing_direction'] = 'left' if tree['default_left'] else 'right'
+                node['missing_type'] = tree['missing_type']
+                node['internal_value'] = tree['internal_value']
+                node['weight'] = tree['internal_weight']
+                node['count'] = tree['internal_count']
+            else:
+                node['leaf_value'] = tree['leaf_value']
```

Comment:
Maybe merge `leaf_value` and `internal_value` into simple `value` as it was done for `count` and `weight`?

Reply:
Hi. I can do this, but I have a couple concerns. The first is that I don't know what `internal_value` represents. I originally excluded it from the output of this method because I didn't think/know if it would be useful. Second, users will intuitively be able to interpret `leaf_value` column as the terminal node predictions for each tree, so they might be confused to find a single `value` column for every node and have trouble interpreting its meaning.

Reply:
https://github.com/microsoft/LightGBM/blob/d7f8aa53b8be5212887e97e474c90bc5d8beac19/include/LightGBM/tree.h#L395-L396
https://github.com/microsoft/LightGBM/blob/d7f8aa53b8be5212887e97e474c90bc5d8beac19/include/LightGBM/tree.h#L432-L434
I suppose that it is `leaf_value` from previous boosting stages. So I guess it should be merged, because it represents the same thing and now confuses even more with significant number of NaN values in the corresponding columns. With `S`/`L` encoding in `node_index` field I think that users will easily find actual leaf output values.

Reply:
I have combined `leaf_value` and `internal_value` to be `value`. However, it seems like `internal_value` is on a different scale than `leaf_value`. Here's a screenshot of the output from the breast cancer dataset for a single tree model with `max_depth = 2`.

![image](https://user-images.githubusercontent.com/7775219/70870808-f7eec600-1f65-11ea-86d7-7e4d7816dbee.png)